File: frontend/lib/models/chat_model.dart
class ChatModel {
  final String chatId;
  final bool isGroup;
  final String contactId;
  final String contactName;
  final String? contactAvatar;
  final int? memberCount;
  final String? lastMessage;
  final String? lastMessageType;
  final DateTime? lastMessageTime;
  final String? lastMessageSenderId;
  final int unreadCount;
  bool isArchived;
  bool isMuted;

  ChatModel({
    required this.chatId,
    this.isGroup = false,
    this.contactId = '',
    required this.contactName,
    this.contactAvatar,
    this.memberCount,
    this.lastMessage,
    this.lastMessageType,
    this.lastMessageTime,
    this.lastMessageSenderId,
    this.unreadCount = 0,
    this.isArchived = false,
    this.isMuted = false,
  });

  factory ChatModel.fromJson(Map<String, dynamic> json) {
    return ChatModel(
      chatId: json['chat_id'] ?? '',
      isGroup: json['is_group'] ?? false,
      contactId: (json['contact_id'] ?? '').toString(),
      contactName: json['contact_username'] ?? 'User',
      contactAvatar: json['contact_avatar'],
      memberCount: json['member_count'] is int
          ? json['member_count']
          : int.tryParse(json['member_count']?.toString() ?? ''),
      lastMessage: json['last_message_content'],
      lastMessageType: json['last_message_type'],
      lastMessageTime: json['last_message_time'] != null
          ? DateTime.parse(json['last_message_time'])
          : null,
      lastMessageSenderId: json['last_message_sender_id'],
      unreadCount: json['unread_count'] is int
          ? json['unread_count']
          : int.tryParse(json['unread_count']?.toString() ?? '') ?? 0,
      isArchived: json['is_archived'] ?? false,
      isMuted: json['is_muted'] ?? false,
    );
  }
}

File: backend/src/models/chat.model.ts
export interface Chat {
    id: string;
    is_group: boolean;
    name?: string | null; // Decrypted group name; NULL for 1:1 chats
    created_by?: string | null;
    created_at: Date;
}

export type ChatParticipantRole = 'owner' | 'member';

export interface ChatParticipant {
    chat_id: string;
    user_id: string;
    role: ChatParticipantRole;
    is_archived: boolean;
    is_muted: boolean;
    is_deleted: boolean;
    cleared_at: Date | null;
}

// A single member row for the "Group Info" members list.
export interface GroupMember {
    user_id: string;
    username: string;
    avatar_url: string | null;
    role: ChatParticipantRole;
}

export interface ChatListItem {
    chat_id: string;
    is_group: boolean;
    contact_id: string | null; // NULL for group chats
    contact_username: string; // The other user's username for 1:1 chats, or the group name for group chats
    contact_avatar?: string | null; // Always NULL for group chats (no group avatar image)
    member_count?: number | null; // Only set for group chats
    is_archived: boolean;
    is_muted: boolean;
    last_message_id?: string | null;
    last_message_content?: string | null; // Decrypted content preview
    last_message_type?: string | null;
    last_message_status?: string | null;
    last_message_sender_id?: string | null;
    last_message_time?: Date | null;
    unread_count: number;
}